src/main/java/tigase/server/rtbl/RTBLRepository.java

```diff
@@ -0,0 +1,459 @@
+		if (rtbl != null) {
+			switch (event.getAction()) {
+				case add -> {
+					rtbl.getBlocked().add(event.getId());
+				}
+				case remove -> {
+					rtbl.getBlocked().remove(event.getId());
+				}
+			}
+		}
+	}
+
+	public void purge(BareJID pubsubJid, String node) {
+		Key key = new Key(pubsubJid, node);
+		RTBL rtbl = cache.get(key);
+		if (rtbl != null) {
+			Set<String> toRemove = new HashSet<>(rtbl.getBlocked());
+			rtbl.getBlocked().clear();
+			for (String item : toRemove) {
+				updateStore(key, Action.remove, item);
+			}
+			eventBus.fire(new RTBLReload(pubsubJid, node));
+		}
+	}
+
+	public void reload(BareJID pubsubJid, String node) {
+		eventBus.fire(new RTBLRemoved(pubsubJid, node));
+	}
+
+	@HandleEvent
+	public void handleReload(RTBLReload event) {
+		try {
+			RTBL rtbl = loadList(event.getKey());
+			if (rtbl != null) {
+				cache.replace(rtbl.getKey(), rtbl);
+			}
+		} catch (TigaseDBException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	private void updateStore(Key key, Action action, String id) {
+		switch (action) {
+			case add -> {
+				try {
+					userRepository.setData(repoUser, key.getSubnode(), id, id);
+				} catch (TigaseDBException ex) {
+					logger.log(Level.WARNING, "failed to save updated RTBL " + key + ", adding " + id + " failed", ex);
+				}
+			}
+			case remove -> {
+				try {
+					userRepository.removeData(repoUser, key.getSubnode(), id);
+				} catch (TigaseDBException ex) {
+					logger.log(Level.WARNING, "failed to save updated RTBL " + key + ", removing " + id + " failed", ex);
+				}
+			}
+		}
+	}
+	
+	private RTBL loadList(Key key) throws TigaseDBException {
+		if (userRepository.userExists(repoUser)) {
+			try {
+				Map<String, String> data = userRepository.getDataMap(repoUser, key.getSubnode());
+				BareJID jid = BareJID.bareJIDInstanceNS(data.remove("jid"));
+				String node = data.remove("node");
+				String hash = data.remove("hash");
+				if (jid == null || hash == null || node == null) {
+					return null;
+				}
+				return new RTBL(jid, node, hash, new CopyOnWriteArraySet<>(data.keySet()));
+			} catch (TigaseDBException ex) {
+				logger.log(Level.WARNING, "could not load RTBL for " + key, ex);
+			}
+		}
+		return null;
+	}
+
+	protected void reload() {
+		try {
+			if (userRepository.userExists(repoUser)) {
+				String[] subnodes = userRepository.getSubnodes(repoUser);
+				if (subnodes != null) {
+					for (String subnode : subnodes) {
+						Key key = Key.parse(subnode);
+						try {
+							if (key != null) {
+								RTBL rtbl = loadList(key);
+								if (rtbl != null) {
+									cache.put(key, rtbl);
+								} else {
+									cache.remove(key);
+								}
+							}
+						} catch (TigaseDBException ex) {
+							logger.log(Level.WARNING, "could not load RTBL " + key, ex);
+						}
+					}
+				}
+			}
+		} catch (TigaseDBException ex) {
+			logger.log(Level.WARNING, "could not load list of RTBLs", ex);
+		}
+	}
+
+	public static class Key {
```

Comment:
Maybe `record`?

Reply:
This most likely could be record, but I do not see a benefit of that, than just to use new shiny thing..

Reply:
Less code - improved readability?